src/map/mob.c

```diff
@@ -2710,10 +2710,20 @@ static int mob_dead(struct mob_data *md, struct block_list *src, int type)
 				drop_rate = (int)(drop_rate*1.25); // pk_mode increase drops if 20 level difference [Valaris]
 
 			// Increase drop rate if user has SC_CASH_RECEIVEITEM
-			if (sd && sd->sc.data[SC_CASH_RECEIVEITEM]) // now rig the drop rate to never be over 90% unless it is originally >90%.
-				drop_rate = max(drop_rate, cap_value((int)(0.5 + drop_rate * (sd->sc.data[SC_CASH_RECEIVEITEM]->val1) / 100.), 0, 9000));
-			if (sd && sd->sc.data[SC_OVERLAPEXPUP])
-				drop_rate = max(drop_rate, cap_value((int)(0.5 + drop_rate * (sd->sc.data[SC_OVERLAPEXPUP]->val2) / 100.), 0, 9000));
+			if (sd) {
+				temp = 0;
+
+				if (src)
+					temp += sd->dropaddrace[md->status.race] + (is_boss(src) ? sd->dropaddrace[RC_BOSS] : sd->dropaddrace[RC_NONBOSS]);
+
+				if (sd->sc.data[SC_CASH_RECEIVEITEM]) // now rig the drop rate to never be over 90% unless it is originally >90%.
+					temp += sd->sc.data[SC_CASH_RECEIVEITEM]->val1;
+
+				if (sd->sc.data[SC_OVERLAPEXPUP])
+					temp += sd->sc.data[SC_OVERLAPEXPUP]->val2;
+
+				drop_rate = max(drop_rate, cap_value((int)(0.5 + drop_rate * temp / 100.), 0, 9000));
```

Comment:
The way the various bonuses stack is changing a bit, is it intentional?

Before this commit, `SC_CASH_RECEIVEITEM` and `SC_OVERLAPEXPUP` were stacking multiplicatively, while now they're stacking additively.

For example, let's assume an item with 10% drop rate (`drop_rate = 1000`), `SC_CASH_RECEIVEITEM` enabled with a +50% bonus (`sc.data[SC_CASH_RECEIVEITEM]->val1 = 150`) and `SC_OVERLAPEXPUP` enabled with a +20% bonus (`sc.data[SC_OVERLAPEXPUP]->val2 = 120`).

Let's put aside the fact that `Malang_Cat_Can` looks wrong, or at least not compatible with the current implementation (instead of +20%, it has a -80% bonus, which might be what this commit is trying to address behind the scenes)

Before this change:
```
if (sd && sd->sc.data[SC_CASH_RECEIVEITEM])
	drop_rate = max(1000, cap_value((int)(0.5 + 1000 * (150) / 100.), 0, 9000)); // => 0.5 + 1000 * 150 / 100 => drop_rate = 1500
if (sd && sd->sc.data[SC_OVERLAPEXPUP])
	drop_rate = max(1500, cap_value((int)(0.5 + 1500 * (120) / 100.), 0, 9000)); // => 0.5 + 1500 * 120 / 100 => drop_rate = 1800

// Final drop rate is 18% = 10% * 150% * 120%
```

After this change:

```
if (sd->sc.data[SC_CASH_RECEIVEITEM])
	temp += 150; // temp = 150
if (sd->sc.data[SC_OVERLAPEXPUP])
	temp += 120; // temp = 270
drop_rate = max(1000, cap_value((int)(0.5 + 1000 * 270 / 100.), 0, 9000)); // => 0.5 + 1000 * 270 / 100 => 2700

// Final drop rate is 27% - this is higher than the expected 17% = 10% * (100% + 50% + 20%) because the base rate of 100 is added in each step rather than just once, resulting in 27% = 10% * (150% + 120%)
```

If the bonuses are meant to stack additively that's fine, but please split it into a separate commit and mention it in the commit message, since it's changing the behavior of existing SCs (and please make sure that it's not adding the base 100% more than once)

Reply:
oh yeah. I coded it way different from the original. Yeah i might revert it back to the original where i shouldn't stack.

Reply:
After extensive thinking and testing, i need to do the whole block regarding drop rate to make the bonuses work properly.